### 实例化一个小型的卷积神经网络

In [11]:
from keras import layers
from keras import models

model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPool2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPool2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))
model.summary()

W0814 17:52:35.741519 140736666145728 deprecation.py:506] From /Users/schp/miniconda2/envs/keras2.0.8/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:1205: calling reduce_prod_v1 (from tensorflow.python.ops.math_ops) with keep_dims is deprecated and will be removed in a future version.
Instructions for updating:
keep_dims is deprecated, use keepdims instead


_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_7 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_5 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_8 (Conv2D)            (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_6 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_9 (Conv2D)            (None, 3, 3, 64)          36928     
_________________________________________________________________
flatten_1 (Flatten)          (None, 576)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 64)                36928     
__________

如上看到的，第一层卷积完了之后，输出为26*26*32，卷积后是高和宽都变窄了，损失了两个像素值。

### 利用卷积网络训练MINST

In [14]:
from keras.datasets import mnist
from keras.utils import to_categorical

(train_images, train_labels), (test_images, test_lables) = mnist.load_data()

train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

train_labels = to_categorical(train_labels)
test_lables = to_categorical(test_lables)

model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])
model.fit(train_images, train_labels, epochs=5, batch_size=64)

W0814 17:59:30.608500 140736666145728 deprecation_wrapper.py:119] From /Users/schp/miniconda2/envs/keras2.0.8/lib/python3.6/site-packages/keras/optimizers.py:711: The name tf.train.Optimizer is deprecated. Please use tf.compat.v1.train.Optimizer instead.

W0814 17:59:30.652453 140736666145728 deprecation.py:506] From /Users/schp/miniconda2/envs/keras2.0.8/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:2755: calling reduce_sum_v1 (from tensorflow.python.ops.math_ops) with keep_dims is deprecated and will be removed in a future version.
Instructions for updating:
keep_dims is deprecated, use keepdims instead
W0814 17:59:30.675905 140736666145728 deprecation_wrapper.py:119] From /Users/schp/miniconda2/envs/keras2.0.8/lib/python3.6/site-packages/keras/backend/tensorflow_backend.py:2759: The name tf.log is deprecated. Please use tf.math.log instead.

W0814 17:59:30.831532 140736666145728 deprecation.py:323] From /Users/schp/miniconda2/envs/keras2.0.8/lib/python3.6/site-pack

Epoch 1/5
60000/60000 [==============================] - 37s - loss: 0.1694 - acc: 0.9462    
Epoch 2/5
60000/60000 [==============================] - 38s - loss: 0.0471 - acc: 0.9852    
Epoch 3/5
60000/60000 [==============================] - 37s - loss: 0.0323 - acc: 0.9905    
Epoch 4/5
60000/60000 [==============================] - 39s - loss: 0.0245 - acc: 0.9924    
Epoch 5/5
60000/60000 [==============================] - 39s - loss: 0.0199 - acc: 0.9939    


In [15]:
test_loss, test_acc = model.evaluate(test_images, test_lables)
test_acc

 9888/10000 [============================>.] - ETA: 0s

0.9923

### 将图像复制到训练、验证和测试的目录

In [1]:
import os
import shutil

original_dataset_dir = '/Users/schp/Desktop/样本/dogs-vs-cats'  # 原始数据集目录
base_dir = '/Users/schp/Desktop/样本/dogs-vs-cats-small'  # 保存较小数据集的目录
os.mkdir(base_dir)

# 划分后的训练、验证和测试的目录
train_dir = os.path.join(base_dir, 'train')
os.mkdir(train_dir)
validation_dir = os.path.join(base_dir, 'validation')
os.mkdir(validation_dir)
test_dir = os.path.join(base_dir, 'test')
os.mkdir(test_dir)

# 猫的训练图像目录
train_cats_dir = os.path.join(train_dir, 'cats')
os.mkdir(train_cats_dir)
# 狗的训练图像目录
train_dogs_dir = os.path.join(train_dir, 'dogs')
os.mkdir(train_dogs_dir)

# 猫的验证图像目录
validation_cats_dir = os.path.join(validation_dir, 'cats')
os.mkdir(validation_cats_dir)
# 狗的验证图像目录
validation_dogs_dir = os.path.join(validation_dir, 'dogs')
os.mkdir(validation_dogs_dir)

# 猫的测试图像目录
test_cats_dir = os.path.join(test_dir, 'cats')
os.mkdir(test_cats_dir)
# 狗的测试图像目录
test_dogs_dir = os.path.join(test_dir, 'dogs')
os.mkdir(test_dogs_dir)

# 复制图像到划分后的训练、验证和测试的目录
# 复制前1000张猫的图像到train_cats_dir
fnames = ['cat.{}.jpg'.format(i) for i in range(1000)]
for fname in fnames:
    src = os.path.join(original_dataset_dir, 'train', fname)
    dst = os.path.join(train_cats_dir, fname)
    shutil.copyfile(src, dst)
    
# 复制500张猫的图像到validation_cats_dir
fnames = ['cat.{}.jpg'.format(i) for i in range(1000, 1500)]
for fname in fnames:
    src = os.path.join(original_dataset_dir, 'train', fname)
    dst = os.path.join(validation_cats_dir, fname)
    shutil.copyfile(src, dst)

# 复制500张猫的图像到test_cats_dir
fnames = ['cat.{}.jpg'.format(i) for i in range(1500, 2000)]
for fname in fnames:
    src = os.path.join(original_dataset_dir, 'train', fname)
    dst = os.path.join(test_cats_dir, fname)
    shutil.copyfile(src, dst)

# 复制前1000张狗的图像到train_cats_dir
fnames = ['dog.{}.jpg'.format(i) for i in range(1000)]
for fname in fnames:
    src = os.path.join(original_dataset_dir, 'train', fname)
    dst = os.path.join(train_dogs_dir, fname)
    shutil.copyfile(src, dst)
    
# 复制500张狗的图像到validation_cats_dir
fnames = ['dog.{}.jpg'.format(i) for i in range(1000, 1500)]
for fname in fnames:
    src = os.path.join(original_dataset_dir, 'train', fname)
    dst = os.path.join(validation_dogs_dir, fname)
    shutil.copyfile(src, dst)

# 复制500张狗的图像到test_cats_dir
fnames = ['dog.{}.jpg'.format(i) for i in range(1500, 2000)]
for fname in fnames:
    src = os.path.join(original_dataset_dir, 'train', fname)
    dst = os.path.join(test_dogs_dir, fname)
    shutil.copyfile(src, dst)

### 检查各目录下图像数量

In [2]:
print('total training cats image:', len(os.listdir(train_cats_dir)))
print('total validation cats image:', len(os.listdir(validation_cats_dir)))
print('total test cats image:', len(os.listdir(test_cats_dir)))

print('total training dogs image:', len(os.listdir(train_dogs_dir)))
print('total validation dogs image:', len(os.listdir(validation_dogs_dir)))
print('total test dogs image:', len(os.listdir(test_dogs_dir)))

total training cats image: 1000
total validation cats image: 500
total test cats image: 500
total training dogs image: 1000
total validation dogs image: 500
total test dogs image: 500


### 构建网络

In [5]:
from keras import layers
from keras import models

model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(150, 150, 3)))  # 150*150随意大小
model.add(layers.MaxPool2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPool2D((2, 2)))
model.add(layers.Conv2D(128, (3, 3), activation='relu'))
model.add(layers.MaxPool2D((2, 2)))
model.add(layers.Conv2D(128, (3, 3), activation='relu'))
model.add(layers.MaxPool2D((2, 2)))
model.add(layers.Flatten())
model.add(layers.Dense(512, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

In [6]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_9 (Conv2D)            (None, 148, 148, 32)      896       
_________________________________________________________________
max_pooling2d_9 (MaxPooling2 (None, 74, 74, 32)        0         
_________________________________________________________________
conv2d_10 (Conv2D)           (None, 72, 72, 64)        18496     
_________________________________________________________________
max_pooling2d_10 (MaxPooling (None, 36, 36, 64)        0         
_________________________________________________________________
conv2d_11 (Conv2D)           (None, 34, 34, 128)       73856     
_________________________________________________________________
max_pooling2d_11 (MaxPooling (None, 17, 17, 128)       0         
_________________________________________________________________
conv2d_12 (Conv2D)           (None, 15, 15, 128)       147584    
__________

### 神经网络参数计算
<font color='red'>如上summary所示：</font>  
<font color='red'>conv2d_9层：</font>  
<font color='red'>原图shape：150\*150\*3</font>  
<font color='red'>filters个数：32</font>  
<font color='red'>卷积核shape：3\*3<font>  
<font color='red'>单个卷积核参数：3\*3\*3\=27<font>  
<font color='red'>conv2d_9层所有卷积核参数：单个卷积核参数\*filters个数。即3\*3\*3\*32=864<font>  
<font color='red'>conv2d_9层所有参数：单个卷积核参数\*filters个数+偏置。即3\*3\*3\*32+32=896<font>  
<font color='red'>max_pooling2d_9层对conv2d_9层的输出进行了缩小，图像shape变为原来的一半，深度变为32<font>   
<font color='red'>conv2d_10层参数计算：3\*3\*32\*64+64=18496<font>  
全连接层dense_5的参数计算：6272\*512+512=3211776  
Total params和Trainable params的参数值为summary打印出的参数之和：896+18496+73856+147584+3211776+513=3453121

